Allow for adding dependencies to a test suites' annotationProcessor configuration in the suite's dependencies block (#20115)
gradle/gradle: subprojects/docs/src/docs/userguide/jvm/jvm_test_suite_plugin.adoc

<1> Configure the built-in `test` test suite.

> <2> Add the assertj library to the test's compile and runtime classpaths.  The `dependencies` block within a test suite is already scoped for that test suite.  Instead of having to know the global name of the configuration, test suites have a consistent name that you use in this block for declaring `implementation`, `compileOnly` and `runtimeOnly` dependencies.

Reviewer: let's add `annotationProcessor` to this too.

The `dependencies` block within a test suite does not (yet) provide access to the full set of dependency configuration methods that are available within a top-level `dependencies` block.

> For instance the link:{groovyDslPath}/org.gradle.api.artifacts.dsl.DependencyHandler.html#org.gradle.api.artifacts.dsl.DependencyHandler:platform(java.lang.Object)[platform()] and link:{groovyDslPath}/org.gradle.api.artifacts.dsl.DependencyHandler.html#org.gradle.api.artifacts.dsl.DependencyHandler:platform(java.lang.Object)[enforcedPlatform()] methods described in the section <<java_platform_plugin#sec:java_platform_bom_import,Sourcing constraints from another platform>> are not yet available.  You will need to use the equivalent way of configuring dependencies noted above to utilize these methods.

Reviewer 2: But it will be pick up if you have defined the `platform` as a general dependency and only the actual artifacts on the suite?

Author: If I understand what you're asking properly, yes.  I just added a test to demonstrate this.

Reviewer 2: Close, but what I meant was if I declared the platform in the regular dependencies block, will this affect the suites as well?
```
dependencies {
  api(platform("..."))
}
```

> The `dependencies` block within a test suite does not (yet) provide access to the full set of dependency configuration methods that are available within a top-level `dependencies` block.
> For instance the link:{groovyDslPath}/org.gradle.api.artifacts.dsl.DependencyHandler.html#org.gradle.api.artifacts.dsl.DependencyHandler:platform(java.lang.Object)[platform()] and link:{groovyDslPath}/org.gradle.api.artifacts.dsl.DependencyHandler.html#org.gradle.api.artifacts.dsl.DependencyHandler:platform(java.lang.Object)[enforcedPlatform()] methods described in the section <<java_platform_plugin#sec:java_platform_bom_import,Sourcing constraints from another platform>> are not yet available.  You will need to use the equivalent way of configuring dependencies noted above to utilize these methods.
> 
> [NOTE]
> ====
> You must access the configurations of a test suite in a top-level `dependencies` block only after creating the test suite in a prior `testing` block.  Adding dependencies is not a lazy operation, and the test suite's configurations must be created prior to adding any dependencies to them using their global names.
> ====
> 

Reviewer: I think this is a little confusing to bring up here because we don't want people to have to use the top-level dependencies block to configure a test suite's dependencies or have to think about converting between the local name and the "global" name in the build script. 

I think the warning about the lack of certain methods in this context is OK, but I would focus this on how to do things through the test suite dependencies block.

I understand the idea behind the note that the order is important. I think calling this out in particular here makes it seem like this is unique or special to test suites, but this is very much the same as any other construct in Gradle.  e.g., if you did this the old way, you wouldn't be able to configure the dependencies before creating the source set. 

```suggestion
The `dependencies` block within a test suite does not provide access to the same methods as the top-level `dependencies` block.

For instance, the link:{groovyDslPath}/org.gradle.api.artifacts.dsl.DependencyHandler.html#org.gradle.api.artifacts.dsl.DependencyHandler:platform(java.lang.Object)[platform()] and link:{groovyDslPath}/org.gradle.api.artifacts.dsl.DependencyHandler.html#org.gradle.api.artifacts.dsl.DependencyHandler:platform(java.lang.Object)[enforcedPlatform()] are not available.

To use these methods, you need to access the top-level dependencies handler directly with `project.dependencies.platform(...)` or `project.dependencies.enforcedPlatform(...)`.
```
